File: p646/p646_2.go
// Sort + Greedy. O(n^2)

type Pair [][]int

func (p Pair) Len() int {
    return len(p)
}

// Greedy!
func (p Pair) Less(i, j int) bool {
    return p[i][1] < p[j][1]
}

func (p Pair) Swap(i, j int) {
    p[i], p[j] = p[j], p[i]
}

func findLongestChain(pairs [][]int) int {
    n := len(pairs)
    if n < 2 {
    	return n
    }

    sort.Sort(Pair(pairs))

    ans, i, end := 0, 0, 0

    for i < n {
        ans++
        end = pairs[i][1]
        
        for i < n && pairs[i][0] <= end {
            i++
        }
    }

    return ans
}

File: p646/p646_2.cpp
// Sort + Greedy. O(N*logN)

bool mycmp(vector<int>& a, vector<int>& b) {
    return a[1] < b[1];
}

class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        int n = pairs.size(), ans = 0;

        sort(pairs.begin(), pairs.end(), mycmp);

        int lastEnd = INT_MIN;
        for (int i = 0; i < n; ++i)
            if (pairs[i][0] > lastEnd)
            {
                ++ans;
                lastEnd = pairs[i][1];
            }
        
        return ans;
    }
};